src/main/scala/optics/poly/EPOptional.scala

```diff
@@ -44,38 +34,49 @@ trait EPOptional[+E, -S, +T, +A, -B] extends EPTraversal[E, S, T, A, B] { self =
       evA.substituteCo[[X] =>> EPOptional[E, S, T, X, B]](this)
     )
 
+  // @alpha("andThen")
+  // def >>>[E1, C, D](other: EPOptional[E1, A, B, C, D]): EPOptional[E | E1, S, T, C, D] = new EPOptional[E | E1, S, T, C, D] {
+  //   def getOrModify(from: S): Either[(E | E1, T), C] =
+  //     for {
+  //       a <- self.getOrModify(from)
+  //       t <- other.getOrModify(a).left.map{ case (e1, b) => (e1, self.replace(b)(from)) }
+  //     } yield t
+  //   override def replace(to: D): S => T = self.modify(other.replace(to))
+  // }
+
 }
 
 object EPOptional {
-  def apply[E, S, T, A, B](_getOrModify: S => Either[(E, T), A], _replace: B => S => T): EPOptional[E, S, T, A, B] = new EPOptional[E, S, T, A, B] {
-    def getOrModify(from: S): Either[(E, T), A] = _getOrModify(from)
-    override def replace(to: B): S => T = _replace(to)
-  }
+  def apply[E, S, T, A, B](_getOrModify: S => Either[(E, T), A], _replaceOrError: B => S => Either[E, T]): EPOptional[E, S, T, A, B] =
+    new EPOptional[E, S, T, A, B] {
+      def getOrModify(from: S): Either[(E, T), A] = _getOrModify(from)
+      override def replaceOrError(to: B): S => Either[E, T] = _replaceOrError(to)
+    }
 }
 
 object POptional {
   def apply[S, T, A, B](_getOrModify: S => Either[T, A], _replace: B => S => T): POptional[S, T, A, B] =
-    EPOptional[Any, S, T, A, B](_getOrModify(_).left.map(defaultError -> _), _replace)
+    EPOptional[Any, S, T, A, B](_getOrModify(_).left.map(defaultError -> _), b => s => Right(_replace(b)(s)))
 }
 
 object EOptional {
-  def apply[Error, From, To](_getOrError: From => Either[Error, To], _replace: To => From => From): EOptional[Error, From, To] =
-    EPOptional(from => _getOrError(from).left.map(_ -> from), _replace)
-
-  def indexMap[K, V](key: K): EOptional[String, Map[K, V], V] =
-    apply[String, Map[K, V], V](
-      _.get(key).toRight(s"Key $key is missing"),
-      newValue => map => if(map.contains(key)) map + (key -> newValue) else map
-    )
-
-  def indexList[A](key: Int): EOptional[String, List[A], A] =
-    apply[String, List[A], A](
-      list => Try(list(key)).toOption.toRight(s"No value at index $key"),
-      newValue => list => Try(list.updated(key, newValue)).getOrElse(list)
-    )
+  def apply[Error, From, To](_getOrError: From => Either[Error, To], _replaceOrError: To => From => Either[Error, From]): EOptional[Error, From, To] =
```

Comment:
I was thinking, for smart constructor `apply`, to only require to define the error in the `get`  part. `replaceOrError` can be implemented using `getOrError` + `replace`. What do you think?